src/lobpcg.jl

```diff
@@ -0,0 +1,751 @@
+    if log
+        return LOBPCGState(iteration, iterator.residuals[1:sizeX], iterator.ritz_values[1:sizeX])
+    else
+        return LOBPCGState(iteration, nothing, nothing)
+    end
+end
+
+"""
+The Locally Optimal Block Preconditioned Conjugate Gradient Method (LOBPCG)
+
+Finds the k extremal eigenvalues and their corresponding eigenvectors satisfying `AX = λBX` returning `λ` and `X`, using `X` as an initial guess, where `k = size(X,2)`. The input `X` is overwritten with the solution.
+
+`A` and `B` may be generic types but `Base.A_mul_B!(C, AorB, X)` and `AorB * X` must be defined for vectors and strided matrices `X` and `C`.
+
+    lobpcg(A, [B,] largest, X0; kwargs...) -> results
+
+# Arguments
+
+- `A`: linear operator;
+- `B`: linear operator;
+- `largest`: `true` if largest eigenvalues are desired and false if smallest;
+- `X0`: Initial guess, will not be modified. The number of columns is the number of eigenvectors desired.
+
+## Keywords
+
+- `log::Bool`: default is `false`; if `true`, `results.trace` will store iterations    
+    states; if `false` only `results.trace` will be empty;
+
+- `P`: preconditioner of residual vectors, must overload `A_ldiv_B!`;
+
+- `C`: constraint to deflate the residual and solution vectors orthogonal
+    to a subspace; must overload `A_mul_B!`;
+
+- `maxiter`: maximum number of iterations; default is 200;
+
+- `tol::Number`: tolerance to which residual vector norms must be under.
+
+# Output
+
+- `results`: a `LOBPCGResults` struct. `r.λ` and `r.X` store the eigenvalues and eigenvectors.
+
+"""
+function lobpcg(A, largest::Bool, X0::Union{AbstractMatrix, AbstractVector}; kwargs...)
+    lobpcg(A, nothing, largest, X0; kwargs...)
+end
+function lobpcg(A, B, largest, X0;
+                log=false, P=nothing, C=nothing, 
+                tol=nothing, maxiter=200)
+
+    X = copy(X0)
+    T = eltype(X)
+    n = size(X, 1)
+    sizeX = size(X, 2)
+    sizeX > n && throw("X column dimension exceeds the row dimension")
+
+    iterator = LOBPCGIterator(A, B, X, largest, P, C)
+    
+    return lobpcg!(iterator, log=log, tol=tol, maxiter=maxiter)
+end
+
+"""
+    lobpcg!(iterator::LOBPCGIterator; kwargs...) -> results
+    
+# Arguments
+
+- `iterator::LOBPCGIterator`: a struct having all the variables required
+    for the LOBPCG algorithm.
+
+## Keywords
+
+- `log::Bool`: default is `false`; if `true`, `results.trace` will store iterations    
+    states; if `false` only `results.trace` will be empty;
+
+- `maxiter`: maximum number of iterations; default is 200;
+
+- `tol::Number`: tolerance to which residual vector norms must be under.
+
+# Output
+
+- `results`: a `LOBPCGResults` struct. `r.λ` and `r.X` store the eigenvalues and eigenvectors.
+
+"""
+function lobpcg!(iterator::LOBPCGIterator; log=false, tol=nothing, maxiter=200)
+    T = eltype(iterator.XBlocks.block)
+    X = iterator.XBlocks.block
+    for j in 1:size(X,2)
+        if all(x -> x==0, view(X, :, j))
+            X[:,j] .= rand.()
```

Comment:
I am curious, why is this working?

Reply:
This is to make sure we don't start with zeros in any of the Ritz vector. Any dependence in the vectors is going to kill the algorithm. I am assuming the random vectors will not be dependent.

Reply:
Can't we just drop this check? Seems wasteful to iterate over the data when it's generally not strictly necessary. Also `rand.()`?

Reply:
This is a small price to pay. I think the most common pre-allocation function is `zeros` not `rand` so without this line, the function will fail more often than not when pre-allocating and the error will not be informative. I can add a flag that turns off this check for the trained user to use.

`rand.()` makes a random vector. `rand()` copies the same random number over all the elements.
```julia
julia> A = zeros(3);

julia> A .= rand()
3-element Array{Float64,1}:
 0.431159
 0.431159
 0.431159

julia> A .= rand.()
3-element Array{Float64,1}:
 0.740519
 0.313979
 0.199173
```

Reply:
:heavy_check_mark: